src/borrow/shared.rs

```diff
@@ -452,10 +447,27 @@ mod tests {
     use crate::untyped_array::PyUntypedArrayMethods;
     use pyo3::ffi::c_str;
 
-    fn get_borrow_flags<'py>(py: Python<'py>) -> &'py BorrowFlagsInner {
+    struct BorrowFlagsState(usize, usize, Option<isize>);
```

Comment:
Very nice solution! I think I would prefer if we gave the fields descriptive names, it looks quite ambiguous and easy to mess up currently. What do you think?